### ISSUE TITLE
rand cluster

### ISSUE THREAD

**Issue by bioguy2018** [pcahan1/singleCellNet#28]
Dear developer,
I have been trying the singlecellnet data on my data, mostly mouse immune cells, which I have processed using scanpy and then converted to a loom file! I have around 12k common genes with two different training dataset and limiting them to certain cell types that I am interested in (10-15 immune cell types available depending on the training set).
Upon the result, most of my cells are assigned to the rand cluster and there is no clear pattern in the heatmap, I get a strong expression for the rand cluster! What can be possibly wrong? do you have any suggestions?  Due to the design, my dataset is supposed to be pretty clear and manually I can identify some clusters very well so I am wondering what might be off during the process?\

p.s. when I try to assess the classifier on my actually data (not the test data) I get the following error 
> Error in `contrasts<-`(`*tmp*`, value = contr.funs[1 + isOF[nn]]) : 
  contrasts can be applied only to factors with 2 or more levels

I couldn't really find what's going wrong as I simply just replaced the parameters with my query data,I would be thankful for your help!

Here is what I do basically:

> stQuery = lfile$sampTab
dim(stQuery) 
expQuery = lfile$expDat
dim(expQuery)
commong = rownames(expQuery)
tm = utils_loadObject("tm10x.rda")
tmexp=tm$expDat
tmst=tm$sampTab

>cts = c("B cell",  "alveolar macrophage","classical monocyte", "early pro-B cell",
        "basophil", "dendritic cell","erythroblast","fibroblast","Fraction A pre-pro B cell",
        "granulocyte", "hematopoietic precursor cell", "late pro-B cell", "granulocytopoietic cell",
        "macrophage", "hepatocyte", "monocyte", "natural killer cell", "T cell", "immature B cell", "immature T cell",
        "late pro-B cell","myeloid cell","promonocyte","proerythroblast","non-classical monocyte",
        "proerythroblast","mast cell", "hematopoietic precursor cell")                               
tmst2 = filter(tmst, cell_ontology_class %in% cts)
tmst2 = droplevels(tmst2)
rownames(tmst2) = as.vector(tmst2$cell) # filter strips rownames

>tmexp2 = tmexp[,rownames(tmst2)]
dim(tmexp2)

>#filter genes
commonGenes = intersect(rownames(tmexp2), commong)
length(commonGenes)
tmexp2_filtered = tmexp2[commonGenes,]

>#training
stList = splitCommon(tmst2, ncells=100, dLevel="cell_ontology_class")
stTrain = stList[[1]]
expTrain = tmexp2_filtered[,rownames(stTrain)]
system.time(class_info2<-scn_train(stTrain = stTrain, expTrain = expTrain,
                                   nTopGenes = 10, nRand = 70, nTrees = 1000,
                                   nTopGenePairs = 25, dLevel = "cell_ontology_class", colName_samp = "cell"))

>#validate data
stTestList = splitCommon(stList[[2]], ncells=100, dLevel="cell_ontology_class") 
stTest = stTestList[[1]]
expTest = tmexp2_filtered[,rownames(stTest)]

>#predict
system.time(classRes_val_all2 <- scn_predict(class_info2[['cnProc']], expTest, nrand = 50))

>#assess
tm_heldoutassessment = assess_comm(ct_scores = classRes_val_all2, stTrain = stTrain, stQuery = stTest,
                                   dLevelSID = "cell", classTrain = "cell_ontology_class",
                                   classQuery = "cell_ontology_class", nRand = 50)
plot_PRs(tm_heldoutassessment)

>#apply to data
system.time(crHS <- scn_predict(class_info2[['cnProc']], expQuery, nrand=50)
tm_assessment = assess_comm(ct_scores = crHS, stTrain = stTrain,
                                 stQuery = stQuery,
                                 classTrain = "cell_ontology_class",
                            classQuery="cluster",dLevelSID ="cell_name")
plot_PRs(tm_pbmc_assessment)
nqRand=50
sgrp = as.vector(stQuery$cluster)
names(sgrp) = as.vector(stQuery$cell_name)
grpRand = rep("rand", nqRand)
names(grpRand) = paste("rand_", 1:nqRand, sep='')
sgrp = append(sgrp, grpRand)
sc_hmClass(crHS, sgrp, max=5000, isBig=TRUE, cCol=F, font=8)
sc_violinClass(sampTab = stQuery,classRes = crHS, sid = "cell_name", dLevel = "cluster", ncol = 12)

**Comment by pcahan1** [comment]
Can you try classifying with unscaled query data? I think it important not
to gene scale your data prior to classification

**Comment by bioguy2018** [comment]
Dear Patrick,
Thanks a lot for your reply!
Do you mean I should use the raw expression? or just skipping the scaling part when I process my data in other word are steps such as normalization and log transformation fine?
I had this question in a previous thread here before that whether I should use raw expression or a processed one would be okay and I was told both are fine.

Thanks a lot!

**Comment by pcahan1** [comment]
We had done some initial analysis comparing raw vs normalized on query and training. It seemed to not make a big difference on the data sets we used then. But, I recommend using only non-gene scaled data or data in which regression has been applied. But yes, you can do the per cell normalization + log transform.

Having said all of this, the fact that your are getting a contrast error is strange. 

How does the classification look on your held out data?

**Comment by bioguy2018** [comment]
Dear Patrick,
I have did try using just normalized expression and I do see much more improvement but I still have many cells in the rand cluster. This raise a question for me because initially I thought that rand cluster is just n random expressions that has been created to have an additional cluster as random or unidentified cells but I can see even though many cells has been assigned to rand cluster according the violin plot but still they have been classified to a cell type and this confused me and I can't understand the rand cluster and relate it to n random expressions as indicated in the paper.
below you can find my classification plot for the query data and the last plot is the result of my held out data! 

p.s. when I draw the violin plot for the classification result of my query data I get the warning n rows contained missing values, is that normal?

![image](https://user-images.githubusercontent.com/46027703/80616384-236bdc00-8a41-11ea-8dcc-a8a849d75755.png)

![image](https://user-images.githubusercontent.com/46027703/80616439-341c5200-8a41-11ea-8071-2f0608674442.png)

![image](https://user-images.githubusercontent.com/46027703/80613722-efdb8280-8a3d-11ea-88ae-96c1fa2e31a4.png)

**Comment by bioguy2018** [comment]
I am also wondering can I skip the rand classification? I know it's a good measurement for assessing the classifier but I already can see the the classifier is predicting >90% of the rand samples correctly as rand (my bar plot). But it also does it for many of other of my cells which is not optimal

**Comment by yuqiyuqitan** [comment]
If your cells are classified as rand, this likely suggests that you might not have the target cell type in your references. As of now, unless you go into adapting the code yourself, you cannot skip creating the rand category in the classifier.
